From 27c64db6dff88ebe9761dfe3b0c073d2a9bf2e41 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Zbigniew=20J=C4=99drzejewski-Szmek?= <zbyszek@in.waw.pl>
Date: Sun, 5 Oct 2014 18:41:31 -0400
Subject: [PATCH] build-sys: use linux/memfd.h if available

linux/memfd.h was added linux 3.17, so it might not be widely
available for a while.

Also, check if memfd_create is defined, for the HAVE_LINUX_MEMFD_H
check to have a chance of succeeding.

Also, collapse all ifdefs for memfd-related stuff, because they
were all added together so there's no need to check separately.
---
 configure.ac         |  2 ++
 src/shared/memfd.c   |  6 +++++-
 src/shared/missing.h | 14 --------------
 3 files changed, 7 insertions(+), 15 deletions(-)

diff --git a/configure.ac b/configure.ac
index e33c8f75ac..caf1f0e54e 100644
--- a/configure.ac
+++ b/configure.ac
@@ -283,6 +283,7 @@ AM_CONDITIONAL([HAVE_PYTHON_DEVEL], [test "$have_python_devel" = "yes"])
 AC_SEARCH_LIBS([dlsym], [dl], [], [AC_MSG_ERROR([*** Dynamic linking loader library not found])])
 AC_CHECK_HEADERS([sys/capability.h], [], [AC_MSG_ERROR([*** POSIX caps headers not found])])
 AC_CHECK_HEADERS([linux/btrfs.h], [], [])
+AC_CHECK_HEADERS([linux/memfd.h], [], [])
 
 # unconditionally pull-in librt with old glibc versions
 AC_SEARCH_LIBS([clock_gettime], [rt], [], [])
@@ -300,6 +301,7 @@ AC_SUBST(RT_LIBS)
 LIBS="$save_LIBS"
 
 AC_CHECK_FUNCS([fanotify_init fanotify_mark])
+AC_CHECK_FUNCS([memfd_create])
 AC_CHECK_FUNCS([__secure_getenv secure_getenv])
 AC_CHECK_DECLS([gettid, pivot_root, name_to_handle_at, setns, LO_FLAGS_PARTSCAN],
                [], [], [[
diff --git a/src/shared/memfd.c b/src/shared/memfd.c
index 1feca0c979..162c12f7a7 100644
--- a/src/shared/memfd.c
+++ b/src/shared/memfd.c
@@ -25,11 +25,15 @@
 #include <sys/mman.h>
 #include <sys/prctl.h>
 
+#ifdef HAVE_LINUX_MEMFD_H
+#  include <linux/memfd.h>
+#endif
+
 #include "util.h"
 #include "bus-label.h"
-#include "missing.h"
 #include "memfd.h"
 #include "utf8.h"
+#include "missing.h"
 
 int memfd_new(const char *name) {
         _cleanup_free_ char *g = NULL;
diff --git a/src/shared/missing.h b/src/shared/missing.h
index 031fe2d1ce..656921d687 100644
--- a/src/shared/missing.h
+++ b/src/shared/missing.h
@@ -65,25 +65,11 @@
 
 #ifndef F_ADD_SEALS
 #define F_ADD_SEALS (F_LINUX_SPECIFIC_BASE + 9)
-#endif
-
-#ifndef F_GET_SEALS
 #define F_GET_SEALS (F_LINUX_SPECIFIC_BASE + 10)
-#endif
 
-#ifndef F_SEAL_SEAL
 #define F_SEAL_SEAL     0x0001  /* prevent further seals from being set */
-#endif
-
-#ifndef F_SEAL_SHRINK
 #define F_SEAL_SHRINK   0x0002  /* prevent file from shrinking */
-#endif
-
-#ifndef F_SEAL_GROW
 #define F_SEAL_GROW     0x0004  /* prevent file from growing */
-#endif
-
-#ifndef F_SEAL_WRITE
 #define F_SEAL_WRITE    0x0008  /* prevent writes */
 #endif
 
